// @fluid-commerce/dam-picker/dist/chunk-3YZTBOS6.cjs
'use strict';

// src/types.ts
var MESSAGE_TYPES = {
  /** Picker app signals it's ready to receive configuration */
  READY: "DAM_PICKER_READY",
  /** SDK sends initialization data (token, filters) */
  INIT: "DAM_PICKER_INIT",
  /** Picker app sends selected asset back to SDK */
  SELECT: "DAM_PICKER_SELECT",
  /** User cancelled the picker */
  CANCEL: "DAM_PICKER_CANCEL",
  /** An error occurred in the picker */
  ERROR: "DAM_PICKER_ERROR",
  /** SDK requests the picker to close */
  CLOSE: "DAM_PICKER_CLOSE"
};
function isPickerMessage(data) {
  if (typeof data !== "object" || data === null) return false;
  const msg = data;
  return typeof msg.type === "string" && Object.values(MESSAGE_TYPES).includes(msg.type);
}
var DEFAULT_PICKER_URL = "https://admin.fluid.app/dam-picker";

// src/iframe-manager.ts
var IframeManager = class {
  config;
  overlay = null;
  iframe = null;
  isVisible = false;
  constructor(config) {
    this.config = config;
  }
  /**
   * Creates the iframe and overlay elements (hidden by default)
   */
  create() {
    if (this.iframe) {
      return this.iframe;
    }
    this.overlay = document.createElement("div");
    this.overlay.setAttribute("data-dam-picker-overlay", "true");
    this.applyOverlayStyles(this.overlay);
    const modal = document.createElement("div");
    this.applyModalStyles(modal);
    this.iframe = document.createElement("iframe");
    this.iframe.setAttribute("data-dam-picker-iframe", "true");
    this.iframe.src = this.config.src;
    this.iframe.allow = "clipboard-write";
    this.applyIframeStyles(this.iframe);
    this.iframe.onload = () => {
      this.config.onLoad?.();
    };
    this.iframe.onerror = () => {
      this.config.onError?.(new Error("Failed to load picker iframe"));
    };
    this.overlay.addEventListener("click", (e) => {
      if (e.target === this.overlay) {
        this.overlay?.dispatchEvent(new CustomEvent("backdropclick"));
      }
    });
    const handleKeydown = (e) => {
      if (e.key === "Escape" && this.isVisible) {
        this.overlay?.dispatchEvent(new CustomEvent("escapekey"));
      }
    };
    document.addEventListener("keydown", handleKeydown);
    this.overlay._keydownHandler = handleKeydown;
    modal.appendChild(this.iframe);
    this.overlay.appendChild(modal);
    document.body.appendChild(this.overlay);
    return this.iframe;
  }
  /**
   * Shows the overlay and iframe
   */
  show() {
    if (!this.overlay) {
      this.create();
    }
    if (this.overlay) {
      this.overlay.style.display = "flex";
      this.overlay.style.opacity = "0";
      void this.overlay.offsetHeight;
      this.overlay.style.opacity = "1";
      this.isVisible = true;
      document.body.style.overflow = "hidden";
    }
  }
  /**
   * Hides the overlay and iframe
   */
  hide() {
    if (this.overlay) {
      this.overlay.style.opacity = "0";
      setTimeout(() => {
        if (this.overlay) {
          this.overlay.style.display = "none";
        }
      }, 200);
      this.isVisible = false;
      document.body.style.overflow = "";
    }
  }
  /**
   * Destroys the iframe and overlay, removing them from the DOM
   */
  destroy() {
    if (this.overlay) {
      const handler = this.overlay._keydownHandler;
      if (handler) {
        document.removeEventListener("keydown", handler);
      }
      this.overlay.remove();
      this.overlay = null;
    }
    this.iframe = null;
    this.isVisible = false;
    document.body.style.overflow = "";
  }
  /**
   * Returns the iframe element
   */
  getIframe() {
    return this.iframe;
  }
  /**
   * Returns the overlay element
   */
  getOverlay() {
    return this.overlay;
  }
  /**
   * Returns whether the picker is currently visible
   */
  getIsVisible() {
    return this.isVisible;
  }
  /**
   * Applies styles to the overlay element
   */
  applyOverlayStyles(overlay) {
    Object.assign(overlay.style, {
      position: "fixed",
      top: "0",
      left: "0",
      right: "0",
      bottom: "0",
      backgroundColor: "rgba(0, 0, 0, 0.5)",
      backdropFilter: "blur(4px)",
      display: "none",
      alignItems: "center",
      justifyContent: "center",
      zIndex: String(this.config.zIndex),
      transition: "opacity 0.2s ease-in-out",
      padding: "16px"
    });
  }
  /**
   * Applies styles to the modal container
   */
  applyModalStyles(modal) {
    Object.assign(modal.style, {
      width: "100%",
      maxWidth: "1200px",
      height: "100%",
      maxHeight: "80vh",
      backgroundColor: "white",
      borderRadius: "12px",
      boxShadow: "0 25px 50px -12px rgba(0, 0, 0, 0.25)",
      overflow: "hidden",
      display: "flex",
      flexDirection: "column"
    });
  }
  /**
   * Applies styles to the iframe element
   */
  applyIframeStyles(iframe) {
    Object.assign(iframe.style, {
      width: "100%",
      height: "100%",
      border: "none",
      backgroundColor: "white"
    });
  }
};

// src/message-bridge.ts
function generateNonce() {
  return Math.random().toString(36).substring(2) + Date.now().toString(36);
}
var MessageBridge = class {
  iframe = null;
  expectedOrigin;
  nonce;
  messageHandler = null;
  // Callbacks
  onReady;
  onSelect;
  onCancel;
  onError;
  constructor(expectedOrigin) {
    this.expectedOrigin = expectedOrigin;
    this.nonce = generateNonce();
  }
  /**
   * Connects the bridge to an iframe and starts listening for messages
   */
  connect(iframe, callbacks) {
    this.iframe = iframe;
    this.onReady = callbacks.onReady;
    this.onSelect = callbacks.onSelect;
    this.onCancel = callbacks.onCancel;
    this.onError = callbacks.onError;
    this.messageHandler = this.handleMessage.bind(this);
    window.addEventListener("message", this.messageHandler);
  }
  /**
   * Disconnects the bridge and stops listening for messages
   */
  disconnect() {
    if (this.messageHandler) {
      window.removeEventListener("message", this.messageHandler);
      this.messageHandler = null;
    }
    this.iframe = null;
  }
  /**
   * Handles incoming postMessage events
   */
  handleMessage(event) {
    if (event.data?.source?.startsWith("react-devtools")) {
      return;
    }
    if (event.data?.type !== MESSAGE_TYPES.READY) {
      if (!this.validateOrigin(event.origin)) {
        return;
      }
    }
    if (!isPickerMessage(event.data)) {
      return;
    }
    const message = event.data;
    switch (message.type) {
      case MESSAGE_TYPES.READY:
        this.onReady?.();
        break;
      case MESSAGE_TYPES.SELECT:
        if (message.payload?.asset) {
          this.onSelect?.(message.payload.asset);
        }
        break;
      case MESSAGE_TYPES.CANCEL:
        this.onCancel?.();
        break;
      case MESSAGE_TYPES.ERROR:
        if (message.payload?.error) {
          this.onError?.(message.payload.error);
        }
        break;
    }
  }
  /**
   * Validates that the message origin matches expected origin
   */
  validateOrigin(origin) {
    if (origin.includes("localhost") || origin.includes("127.0.0.1")) {
      return true;
    }
    try {
      const expectedUrl = new URL(this.expectedOrigin);
      const actualUrl = new URL(origin);
      return expectedUrl.hostname === actualUrl.hostname;
    } catch {
      return false;
    }
  }
  /**
   * Sends INIT message to the picker iframe
   */
  sendInit(token, filters) {
    const message = {
      type: MESSAGE_TYPES.INIT,
      nonce: this.nonce,
      payload: {
        token,
        filters
      }
    };
    this.postMessage(message);
  }
  /**
   * Sends CLOSE message to the picker iframe
   */
  sendClose() {
    const message = {
      type: MESSAGE_TYPES.CLOSE,
      nonce: this.nonce
    };
    this.postMessage(message);
  }
  /**
   * Posts a message to the iframe
   */
  postMessage(message) {
    if (!this.iframe?.contentWindow) {
      console.warn("[DamPicker] Cannot send message: iframe not available");
      return;
    }
    this.iframe.contentWindow.postMessage(message, this.expectedOrigin);
  }
};

// src/dam-picker.ts
var DEFAULTS = {
  zIndex: 1e4,
  loadTimeout: 1e4
  // 10 seconds
};
var DamPicker = class {
  options;
  iframeManager = null;
  messageBridge = null;
  state = "idle";
  loadTimeoutId = null;
  constructor(options) {
    this.options = {
      zIndex: DEFAULTS.zIndex,
      ...options
    };
    this.validateOptions();
  }
  /**
   * Opens the picker modal
   */
  open() {
    if (this.state === "open" || this.state === "loading") {
      console.warn("[DamPicker] Picker is already open or loading");
      return;
    }
    this.state = "loading";
    const pickerUrl = this.getPickerUrl();
    this.iframeManager = new IframeManager({
      src: pickerUrl,
      zIndex: this.options.zIndex ?? DEFAULTS.zIndex,
      onLoad: () => {
      },
      onError: (error) => {
        this.handleError({
          code: "IFRAME_LOAD_ERROR",
          message: error.message
        });
      }
    });
    this.messageBridge = new MessageBridge(pickerUrl);
    const iframe = this.iframeManager.create();
    this.messageBridge.connect(iframe, {
      onReady: () => this.handleReady(),
      onSelect: (asset) => this.handleSelect(asset),
      onCancel: () => this.handleCancel(),
      onError: (error) => this.handleError(error)
    });
    const overlay = this.iframeManager.getOverlay();
    if (overlay) {
      overlay.addEventListener("backdropclick", () => this.handleCancel());
      overlay.addEventListener("escapekey", () => this.handleCancel());
    }
    this.iframeManager.show();
    this.loadTimeoutId = setTimeout(() => {
      if (this.state === "loading") {
        this.handleError({
          code: "TIMEOUT",
          message: "Picker failed to load within timeout period"
        });
      }
    }, DEFAULTS.loadTimeout);
  }
  /**
   * Closes the picker modal
   */
  close() {
    this.clearLoadTimeout();
    this.messageBridge?.sendClose();
    this.iframeManager?.hide();
    this.state = "closed";
  }
  /**
   * Destroys the picker instance, cleaning up all resources
   */
  destroy() {
    this.clearLoadTimeout();
    this.messageBridge?.disconnect();
    this.messageBridge = null;
    this.iframeManager?.destroy();
    this.iframeManager = null;
    this.state = "idle";
  }
  /**
   * Returns whether the picker is currently open
   */
  isOpen() {
    return this.state === "open" || this.state === "loading";
  }
  /**
   * Returns the current state of the picker
   */
  getState() {
    return this.state;
  }
  /**
   * Handles the READY message from the picker iframe
   */
  handleReady() {
    this.clearLoadTimeout();
    this.state = "ready";
    this.messageBridge?.sendInit(this.options.token, this.options.filters);
    this.state = "open";
    this.options.onReady?.();
  }
  /**
   * Handles asset selection from the picker
   */
  handleSelect(asset) {
    this.close();
    this.options.onSelect(asset);
  }
  /**
   * Handles cancel/close from the picker
   */
  handleCancel() {
    this.close();
    this.options.onCancel?.();
  }
  /**
   * Handles errors from the picker
   */
  handleError(error) {
    this.close();
    this.options.onError?.(error);
  }
  /**
   * Gets the picker app URL based on configuration
   */
  getPickerUrl() {
    return this.options.baseUrl ?? DEFAULT_PICKER_URL;
  }
  /**
   * Clears the load timeout if set
   */
  clearLoadTimeout() {
    if (this.loadTimeoutId) {
      clearTimeout(this.loadTimeoutId);
      this.loadTimeoutId = null;
    }
  }
  /**
   * Validates the constructor options
   */
  validateOptions() {
    if (!this.options.token) {
      throw new Error("[DamPicker] token is required");
    }
    if (typeof this.options.onSelect !== "function") {
      throw new Error("[DamPicker] onSelect callback is required");
    }
  }
};

exports.DEFAULT_PICKER_URL = DEFAULT_PICKER_URL;
exports.DamPicker = DamPicker;
exports.MESSAGE_TYPES = MESSAGE_TYPES;
