// @fluid-commerce/dam-picker/dist/react.js
import { DamPicker } from './chunk-OZXU3554.js';
export { DamPicker } from './chunk-OZXU3554.js';
import { useRef, useState, useCallback, useEffect } from 'react';

function useDamPicker(config) {
  const pickerRef = useRef(null);
  const [isOpen, setIsOpen] = useState(false);
  const configRef = useRef(config);
  configRef.current = config;
  const open = useCallback((callbacks) => {
    pickerRef.current = new DamPicker({
      ...configRef.current,
      onSelect: (asset) => {
        setIsOpen(false);
        callbacks.onSelect(asset);
      },
      onCancel: () => {
        setIsOpen(false);
        callbacks.onCancel?.();
      },
      onError: (error) => {
        setIsOpen(false);
        callbacks.onError?.(error);
      },
      onReady: callbacks.onReady
    });
    setIsOpen(true);
    pickerRef.current.open();
  }, []);
  const close = useCallback(() => {
    pickerRef.current?.close();
    setIsOpen(false);
  }, []);
  useEffect(() => {
    return () => {
      pickerRef.current?.destroy();
      pickerRef.current = null;
    };
  }, []);
  return {
    open,
    close,
    isOpen
  };
}
function DamPickerModal({
  isOpen,
  token,
  onSelect,
  onCancel,
  onError,
  onReady,
  baseUrl,
  zIndex,
  filters
}) {
  const pickerRef = useRef(null);
  const wasOpenRef = useRef(false);
  const callbacksRef = useRef({ onSelect, onCancel, onError, onReady });
  callbacksRef.current = { onSelect, onCancel, onError, onReady };
  useEffect(() => {
    const wasOpen = wasOpenRef.current;
    wasOpenRef.current = isOpen;
    if (isOpen && !wasOpen) {
      pickerRef.current = new DamPicker({
        token,
        baseUrl,
        zIndex,
        filters,
        onSelect: (asset) => callbacksRef.current.onSelect(asset),
        onCancel: () => callbacksRef.current.onCancel?.(),
        onError: (error) => callbacksRef.current.onError?.(error),
        onReady: () => callbacksRef.current.onReady?.()
      });
      pickerRef.current.open();
    } else if (!isOpen && wasOpen) {
      pickerRef.current?.destroy();
      pickerRef.current = null;
    }
  }, [isOpen, token, baseUrl, zIndex, filters]);
  useEffect(() => {
    return () => {
      pickerRef.current?.destroy();
      pickerRef.current = null;
    };
  }, []);
  return null;
}

export { DamPickerModal, useDamPicker };
