// @fluid-commerce/dam-picker/dist/react.cjs
'use strict';

var chunk3YZTBOS6_cjs = require('./chunk-3YZTBOS6.cjs');
var react = require('react');

function useDamPicker(config) {
  const pickerRef = react.useRef(null);
  const [isOpen, setIsOpen] = react.useState(false);
  const configRef = react.useRef(config);
  configRef.current = config;
  const open = react.useCallback((callbacks) => {
    pickerRef.current = new chunk3YZTBOS6_cjs.DamPicker({
      ...configRef.current,
      onSelect: (asset) => {
        setIsOpen(false);
        callbacks.onSelect(asset);
      },
      onCancel: () => {
        setIsOpen(false);
        callbacks.onCancel?.();
      },
      onError: (error) => {
        setIsOpen(false);
        callbacks.onError?.(error);
      },
      onReady: callbacks.onReady
    });
    setIsOpen(true);
    pickerRef.current.open();
  }, []);
  const close = react.useCallback(() => {
    pickerRef.current?.close();
    setIsOpen(false);
  }, []);
  react.useEffect(() => {
    return () => {
      pickerRef.current?.destroy();
      pickerRef.current = null;
    };
  }, []);
  return {
    open,
    close,
    isOpen
  };
}
function DamPickerModal({
  isOpen,
  token,
  onSelect,
  onCancel,
  onError,
  onReady,
  baseUrl,
  zIndex,
  filters
}) {
  const pickerRef = react.useRef(null);
  const wasOpenRef = react.useRef(false);
  const callbacksRef = react.useRef({ onSelect, onCancel, onError, onReady });
  callbacksRef.current = { onSelect, onCancel, onError, onReady };
  react.useEffect(() => {
    const wasOpen = wasOpenRef.current;
    wasOpenRef.current = isOpen;
    if (isOpen && !wasOpen) {
      pickerRef.current = new chunk3YZTBOS6_cjs.DamPicker({
        token,
        baseUrl,
        zIndex,
        filters,
        onSelect: (asset) => callbacksRef.current.onSelect(asset),
        onCancel: () => callbacksRef.current.onCancel?.(),
        onError: (error) => callbacksRef.current.onError?.(error),
        onReady: () => callbacksRef.current.onReady?.()
      });
      pickerRef.current.open();
    } else if (!isOpen && wasOpen) {
      pickerRef.current?.destroy();
      pickerRef.current = null;
    }
  }, [isOpen, token, baseUrl, zIndex, filters]);
  react.useEffect(() => {
    return () => {
      pickerRef.current?.destroy();
      pickerRef.current = null;
    };
  }, []);
  return null;
}

Object.defineProperty(exports, "DamPicker", {
  enumerable: true,
  get: function () { return chunk3YZTBOS6_cjs.DamPicker; }
});
exports.DamPickerModal = DamPickerModal;
exports.useDamPicker = useDamPicker;
